tools/user_migration/migrate_user.py

```diff
@@ -0,0 +1,288 @@
+    is now on Spanner
+
+    """
+    cursor = databases['token'].cursor()
+    cursor.execute(
+        """
+        UPDATE
+            users
+        SET
+            replaced_at = {timestamp},
+            nodeid = {nodeid}
+        WHERE
+           uid = {uid}
+        """.format(
+            timestamp=int(time.time() * 100),
+            nodeid=SPANNER_NODE_ID,
+            uid=user)
+        )
+
+
+# The following two functions are taken from browserid.utils
+def encode_bytes_b64(value):
+    return base64.urlsafe_b64encode(value).rstrip(b'=').decode('ascii')
+
+
+def format_key_id(keys_changed_at, key_hash):
+    return "{:013d}-{}".format(
+        keys_changed_at,
+        encode_bytes_b64(key_hash),
+    )
+
+
+def get_fxa_id(databases, user):
+    """generate the spanner user key values from the original storage
+    data.
+
+    """
+    sql = """
+        SELECT
+            email, generation, keys_changed_at, client_state
+        FROM users
+            WHERE uid = {uid}
+    """.format(uid=user)
+    cursor = databases['mysql'].cursor()
+    cursor.execute(sql)
+    (email, generation, keys_changed_at, client_state) = cursor.next()
+    fxa_uid = email.split('@')[0]
+    fxa_kid = format_key_id(
+        keys_changed_at or generation,
+        bytes.fromhex(client_state),
+    )
+    cursor.close()
+    return (fxa_kid, fxa_uid)
+
+
+def dumper(columns, values):
+    """verbose column and data dumper. """
+    for row in values:
+        print("---\n")
+        for i in range(0, len(columns)):
+            print("{} => {}".format(columns[i], row[i]))
+    print("===\n")
+
+
+def move_user(databases, user, args):
+    """copy user info from original storage to new storage."""
+    # bso column mapping:
+    # id => bso_id
+    # collection => collection_id
+    # sortindex => sortindex
+    # modified => modified
+    # payload => payload
+    # payload_size => NONE
+    # ttl => expiry
+
+    # user collections require a unique key.
+    unique_key_filter = []
+
+    uc_columns = (
+        'fxa_kid',
+        'fxa_uid',
+        'collection_id',
+        'modified',
+    )
+    bso_columns = (
+            'collection_id',
+            'fxa_kid',
+            'fxa_uid',
+            'bso_id',
+            'expiry',
+            'modified',
+            'payload',
+            'sortindex',
+    )
+
+    # Genereate the Spanner Keys we'll need.
+    (fxa_kid, fxa_uid) = get_fxa_id(databases, user)
+
+    # Fetch the BSO data from the original storage.
+    sql = """
+    SELECT collection, id, ttl, modified, payload, sortindex
+    FROM bso
+    WHERE userid = {}
+    ORDER BY modified DESC"""
+
+    count = 0
+    with databases['spanner'].batch() as batch:
+        cursor = databases['mysql'].cursor()
+        try:
+            cursor.execute(sql.format(user))
+            print("Processing... {} -> {}:{}".format(
+                user, fxa_uid, fxa_kid))
+            for (cid, bid, exp, mod, pay, sid) in cursor:
```

Comment:
with a large enough set of bsos, the spanner `batch()` could run into the mutation limit :(

Reply:
Really hoping that since this is per-user based, a user doesn't have *that* many BSOs, but yeah. I'm debating switching this to use a `transaction()` internal call, but that seems equally gross.